components/common-go/baseserver/server.go

```diff
@@ -340,18 +269,121 @@ func (s *Server) initializeGRPC() error {
+			cfg.TLS.CA, cfg.TLS.Cert, cfg.TLS.Key,
+			common_grpc.WithSetClientCAs(true),
+			common_grpc.WithServerName(s.Name),
+		)
+		if err != nil {
+			log.WithError(err).Fatal("cannot load ws-manager certs")
+		}
+
+		opts = append(opts, grpc.Creds(credentials.NewTLS(tlsConfig)))
+	}
+
+	s.grpc = grpc.NewServer(opts...)
 
 	// Register health service by default
-	grpc_health_v1.RegisterHealthServer(s.grpc, s.cfg.grpcHealthCheck)
+	grpc_health_v1.RegisterHealthServer(s.grpc, s.options.grpcHealthCheck)
 
 	return nil
 }
 
-func (s *Server) isGRPCEnabled() bool {
-	return s.cfg.grpcPort >= 0
+func httpAddress(cfg *ServerConfiguration, l net.Listener) string {
+	if l == nil {
+		return ""
+	}
+	protocol := "http"
+	if cfg != nil && cfg.TLS != nil {
+		protocol = "https"
+	}
+	return fmt.Sprintf("%s://%s", protocol, l.Addr().String())
+}
+
+func (s *Server) DebugAddress() string {
+	if s.builtinServices == nil {
+		return ""
+	}
+	return "http://" + s.builtinServices.Debug.Addr
+}
+func (s *Server) HealthAddr() string {
+	if s.builtinServices == nil {
+		return ""
+	}
+	return "http://" + s.builtinServices.Health.Addr
+}
+func (s *Server) HTTPAddress() string {
+	return httpAddress(s.options.config.Services.HTTP, s.httpListener)
+}
+func (s *Server) GRPCAddress() string { return s.options.config.Services.GRPC.GetAddress() }
+
+const (
+	BuiltinDebugPort   = 6060
+	BuiltinMetricsPort = 9500
+	BuiltinHealthPort  = 9501
+)
+
+type builtinServices struct {
+	underTest bool
+
+	Debug   *http.Server
+	Health  *http.Server
+	Metrics *http.Server
+}
+
+func newBuiltinServices(server *Server) *builtinServices {
+	healthAddr := fmt.Sprintf(":%d", BuiltinHealthPort)
+	if server.options.underTest {
+		healthAddr = "localhost:0"
+	}
+
+	return &builtinServices{
+		underTest: server.options.underTest,
+		Debug: &http.Server{
+			Addr:    fmt.Sprintf("localhost:%d", BuiltinDebugPort),
+			Handler: pprof.Handler(),
+		},
+		Health: &http.Server{
+			Addr:    healthAddr,
+			Handler: server.healthEndpoint(),
+		},
+		Metrics: &http.Server{
+			Addr:    fmt.Sprintf("localhost:%d", BuiltinMetricsPort),
+			Handler: server.metricsEndpoint(),
+		},
+	}
+}
+
+func (s *builtinServices) ListenAndServe() error {
+	if s == nil {
+		return nil
+	}
+
+	var eg errgroup.Group
+	if !s.underTest {
+		eg.Go(func() error { return s.Debug.ListenAndServe() })
+		eg.Go(func() error { return s.Metrics.ListenAndServe() })
+	}
+	eg.Go(func() error {
+		// health is the only service which has a variable address,
+		// because we need the health service to figure out if the
+		// server started at all
+		l, err := net.Listen("tcp", s.Health.Addr)
+		if err != nil {
+			return err
+		}
+		s.Health.Addr = l.Addr().String()
+		return s.Health.Serve(l)
+	})
+	return eg.Wait()
 }
 
-func (s *Server) isHTTPEnabled() bool {
-	return s.cfg.httpPort >= 0
+func (s *builtinServices) Close() error {
+	var eg errgroup.Group
+	eg.Go(func() error { return s.Debug.Close() })
```

Comment:
Neat! This does however not ensure that the Debug server is closed last. In general, it's beneficial to ensure it's last as it allows you to get the most metrics out of the system before it shuts down, and that way it also does contain metrics from the http/grpc servers shutting down.